Add Array and Dictionary Destructuring to GDScript

joao-pedro-braz opened the pull request:
# Destructuring Expression in GDScript

This PR adds a new expression to GDScript to allow the easy and somewhat safe destructuring of Arrays and Dictionaries. It was initially based off https://github.com/godotengine/godot-proposals/issues/2135 and further discussed over RocketChat.
It's also part of the current GDScript 4.1 Roadmap.

## TODOs

- [x] Create a TODO

## Getting intimate

Although distructuring Arrays is different than distructuring Dictionaries, they can be mixed together and do share some similarities, namely:
- Both can happen over a variable:
  ```gdscript
  var [var a: int] = [123]
  var { foo = var b: int } = { foo = 321 }
  ``` 
- Both can happen over an identifier (referencing an existing variable):
  ```gdscript
  var a: int
  var [a] = [123]
  var b: int
  var { foo = b } = { foo = 321 }
  ``` 
- Both can be mixed together (also called inner destructuring):
  ```gdscript
  var [{ foo = var foo }] = [{ foo = 123 }]
  print(foo) # Will print 123
  var { foo = [var bar] } = { foo = [123] }
  print(bar # Will print 123
  ```

### Type-checking

Type-checking has also been added wherever possible:
- Array destructuring can only happen over Arrays:
  ```gdscript
  var [var foo] = 123 # This will error
  ```
- If conditions are met, type-checking is done between the destructured element and the source array/dictionary:
  ```gdscript
  var [var foo: int] = ["123"] # Will error: Mismatched type "int" at index 0 in the destructuring expression, expected "String".
  var { foo = var foo: int } = { foo = "123" } # Will error: Mismatched type "int" with key "foo" in the destructuring expression, expected "String".
  ```

### Arrays

Getting into the specifics, destructuring for Arrays has the following features:
- Skip over certain indices:
  ```gdscript
  var [_, var foo] = [123, 321]
  print(foo) # Will print 321
  ```
- Collect all non-destructured elements into a single element (also known as a spread):
  ```gdscript
  var [var foo, var ..remaining] = [123, 321, 123321]
  print(foo) # Will print 123
  print(remaining) # Will print [321, 123321]
  ```

### Dictionaries

And destructuring for Dictionaries has the following features:
- Plain identifying keys:
  ```gdscript
  var { foo = var a } = { foo = 123 }
  print(a) # Will print 123
  ```
- Computed identifying keys (currently only supports identifiers, which references existing variables):
  ```gdscript
  var foo_key := "foo"
  var { foo_key: var a } = { foo = 123 }
  print(a) # Will print 123
  ```
- Collect all non-destructured elements into a single element (also known as a spread):
  ```gdscript
  var { foo = var a, var ..b } = { foo = 123, bar = 321, foobar = 123321 }
  print(a) # Will print 123
  print(b) # Will print { "bar": 321, "foobar": 123321 }
  ```

### Real world example

And because a feature is only as good as it's useful, this is an (working) example of an EditorScript that loads some JSON and destructures it using the syntax this PR introduces:
```gdscript
@tool
extends EditorScript


func _run():
	var client := HTTPRequest.new()
	get_editor_interface().get_base_control().add_child(client)
	client.request_completed.connect(
		func(_result: int, _response_code: int, _headers: PackedStringArray, data: PackedByteArray):
			var [_, {
				userId = var user_id: int,
				id = var id: int,
				title = var title: String,
				body = var body: String,
			}] = JSON.parse_string(data.get_string_from_utf8())
			print("Second Post:\nuserId: %s\nid: %s\ntitle: %s\nbody: %s" % [user_id, id, title, body])
	)
	client.request("https://jsonplaceholder.typicode.com/posts")
```
Which prints:
```
Second Post:
userId: 1
id: 2
title: qui est esse
body: est rerum tempore vitae
sequi sint nihil reprehenderit dolor beatae ea dolores neque
fugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis
qui aperiam non debitis possimus qui neque nisi nulla
```

Closes https://github.com/godotengine/godot-proposals/issues/2135

joao-pedro-braz committed:
Add destructuring to GDScript

anvilfolk commented on modules/gdscript/gdscript_compiler.cpp:
```diff
@@ -165,6 +165,173 @@ GDScriptDataType GDScriptCompiler::_gdtype_from_datatype(const GDScriptParser::D
+
+				// Slice the Array
+				codegen.generator->write_call(temp, p_source, R"(slice)", { codegen.add_constant(idx) });
+				if (r_error) {
```
This hasn't been changed since it was passed as a parameter - it feels like something's missing?

joao-pedro-braz replied:
Yeah, I probably just copy-pasted that

joao-pedro-braz commented:
Fixed issues @anvilfolk pointed out and did some improvements:
- Computed key syntax updated from `[ [a] = ... ]` to `[ a: ... ]` (parity with Dictionaries)
- Removed type-checking restrictions on the right-hand side of a Dictionary Destructuring, which now allows destructuring to be done with basically any type (Including things like `Vector2` and custom Classes)

speakk commented:
I love the effort and completeness here, but the syntax is imo not easy to read nor comprehend - and I'm used to both Rust and Typescript destructuring.

Namely this syntax: `var { foo = var b: int } = { foo = 321 }` makes my brain just do a triple take on what I'm reading.

Personally I'd prefer something like `var { b: int = foo } = { foo = 321 }`
You are after all assigning foo into b, and not b into foo. This is the usual and known assignment syntax.
`b = foo` means b will get get the value of foo.


Secondly I might be missing it in your examples, but is there support for the simpler non-aliasing use case of:
`var { foo: int } = { foo = 321 }`

(as in the use case where you don't want to rename the destructured variable)
